### ISSUE TITLE
Allow --line-range to count from the back of the file to act similar to "tail"

### ISSUE THREAD

**Issue by PlasmaHH** [sharkdp/bat#2944]
For python users it would probably best if -r acts somewhat like the python syntax...

So for negative numbers it could count from backwards, so "tail -n 5" would be "bat -r -5:" or so and then one could display the first and last portion of a file like "bat -r :5 -r -5:"

**Comment by ajesipow** [comment]
I would also welcome such a feature and would be happy to contribute a PR for it.

But I think it's not a trivial change and would likely have performance implications that may not be desired.

I see two options:
1. Doing a first quick pass on the input to determine the max line count, then calculating the absolute values in `LineRange` for the negative relative line number accordingly. A second pass would handle lines just as today. No buffering needed.

2. Buffering the input for multiple lines in `Controller::print_file_ranges` instead of the single-line `line_buffer`. The buffer size would need to be equivalent to how far from the end of the input the user wants the output (which could be very large of course - resulting in potentially unexpected memory usage). `LineRange` would also need adapting to represent the relative negative value somehow and this would most likely be a breaking change. 

Also, today a range with a relative negative value such as `10:-3` is already valid and translated into an absolute line range of `7:10`. With the feature discussed here, new users may expect this to translate to "from line 10 to the 3rd to last line" instead.

Alternatively, to not break the existing logic, relative negative numbers may only occur by in an open-ended range, i.e. `-3:` or `:-3`.

**Comment by atimofeev** [comment]
I made this wrapper script, which utilizes relative negative value functionality mentioned above:
```bash
#!/usr/bin/env bash
file="$1"
show="${2:-10}"
lines=$(wc -l < "$file")
range="$lines:-$((show - 1))"

exec bat "$file" --color=always --style=plain --paging=never --line-range "$range"
```
The second argument is optional and by default shows 10 lines.
It even works if a file has fewer lines than a show value.

**Comment by PlasmaHH** [comment]
> > But I think it's not a trivial change and would likely have performance implications that may not be desired.
> 
> Well, there is tail(1), so whatever it does should be applicable here :-)
> 
> > I see two options:
> 
> Note that these options are not equivalent in presence of non-regular files. IOWs, they are not alternatives, but rather both should be implemented with the decision made depending on whether the target file is seekable.

I am not quite sure what being seekable or not has to do with anything, after all to determine the number of lines in usual text files one has to go through them.


Anyways for one way to implement it, at work we have a program that basically reads some binary format and for each "record" outputs one bit of text. For each of these text bits it buffers the last N ones, and once it is through the file outputs only those. 

The same thing could be done here, one line per entry. The thing why I think the above mentioned "quick determination of lines and then start from there" won't really work is because in most cases syntax highlighting will be broken if it doesnt know about a bit of context there.

**Comment by PlasmaHH** [comment]
> However, I wonder how much of a problem it will be in practice?
> 
> File formats that people usually want to `tail` are line-oriented: logs, for instance. And if you are tailing your logs, then you very much _do not_ want to run the entire log through the syntax highlighter — it would be prohibitively slow. This is, in fact, how I discovered this issue: I tried to `bat` a log file and mindlessly hit End to scroll to the last screenful of the output in the pager — only to discover that it took ~2 minutes for bat to render a 30 MiB file!

For such format this is true. Maybe the most sense would be a "syntax dissector" that keeps for the past lines just so much context lines as necessary, and only when the highlighting should be done it will start that many lines back? For logfiles that would be empty, for others possibly more. I right now don't know the ways bat can highlight xml files, but I for those I like a different colour per tag depth, or for C++ etc. I like different colours for different bracked depths. etc. which would then need to keep lines from the last top level (and when you would do semantic highlighting you would need everything really)

**Comment by keith-hall** [comment]
Thanks for moving the discussion forward. 🙂

> 2\. Buffering the input for multiple lines in `Controller::print_file_ranges` instead of the single-line `line_buffer`. The buffer size would need to be equivalent to how far from the end of the input the user wants the output (which could be very large of course - resulting in potentially unexpected memory usage).

I think the increased memory usage for buffering is justified and shouldn't be super excessive for most use cases. Even buffering 10k lines at an average line length of 1k isn't that bad.

> Alternatively, to not break the existing logic, relative negative numbers may only occur in an open-ended range, i.e. `-3:` or `:-3`.

Probably it makes sense to start with this. If we get demand for more changes later we can consider opening it up more then.

--- 

Buffering in all cases regardless of file seekability could help reduce the number of code paths and make maintenance easier. Probably it would be my personal choice.

About highlighting context lines, currently there is no way for `bat` to know which syntaxes need context or how much. But it certainly makes sense to skip highlighting all prior lines of line oriented log file. Perhaps some mapping (even just a line oriented vs non line oriented setting) per syntax could make sense.

**Comment by ajesipow** [comment]
I opened a PR for this, looking forward to feedback.

**Comment by lmmx** [comment]
It was merged but not yet released (current version is 0.25.0 was released in January 2025)

Unfortunately it doesn't actually fulfil the original request, for `bat -r -10: file.rs`

```rust
error: unexpected argument '-1' found

  tip: to pass '-1' as a value, use '-- -1'

Usage: bat [OPTIONS] [FILE]...
       bat <COMMAND>

For more information, try '--help'.
```

Likewise `bat -r "-10:"` doesn't work, it must be passed with `=` as `bat -r="-10:"`

(I suspect this is the limitation of the underlying clap CLI parser and not something that bat can do anything about?)

![Image](https://github.com/user-attachments/assets/80fd8455-d098-4848-9d01-1cf609734a95)

**Comment by keith-hall** [comment]
Yeah I think because the leading `-` of the number after `-r` gets treated as another switch instead of the value, I don't think it's easily fixable. Possible alternative solutions (as opposed to the equals one) I can think of would be to allow some other character as an alias for minus, or have a different switch where the minus is implied...

**Comment by lmmx** [comment]
Or more simply, the builder form of [Arg::allow_leading_hyphen(true)](https://docs.rs/clap/2.29.0/clap/struct.Arg.html#method.allow_hyphen_values) i.e. call `.allow_hyphen_values(true)` on:

https://github.com/sharkdp/bat/blob/83ffd1ea09b539c7fbf01e88264fd7c9c4ecb9e4/src/bin/bat/clap_app.rs#L516-L534

> The fix is to use [`Arg::allow_leading_hyphen(true)`](https://docs.rs/clap/2.29.0/clap/struct.Arg.html#method.allow_hyphen_values) for just `--argument` or globally for the entire program (not usually recommended) with [`AppSettings::AllowLeadingHyphen`](https://docs.rs/clap/2.29.0/clap/enum.ArgSettings.html#variant.AllowLeadingHyphen).

(via https://github.com/clap-rs/clap/issues/1134#issuecomment-356352402)

I’ve opened a small PR #3442 for this (if it is indeed that simple)
